cw-relayer/tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/ojo-network/cw-relayer/tests/e2e/orchestrator"
+	"github.com/ojo-network/cw-relayer/tools"
+)
+
+func (s *IntegrationTestSuite) TestQueryRates() {
+	address := common.HexToAddress(orchestrator.ContractAddress)
+	ethClient, err := ethclient.Dial(s.orchestrator.EVMRpc)
+	s.Require().NoError(err)
+
+	oracle, err := client.NewOracle(address, ethClient)
+	s.Require().NoError(err)
+
+	mockPrices := s.priceServer.GetMockPrices()
+	s.Require().NotZero(len(mockPrices))
+
+	callOpts := bind.CallOpts{
+		Pending: false,
+	}
+
+	session := &client.OracleSession{
+		Contract: oracle,
+		CallOpts: callOpts,
+	}
+
+	// handle delay in deployment of contract
+	checkDenom := tools.StringToByte32(mockPrices[0].Denom)
+	_, err = oracle.GetPriceData(&callOpts, checkDenom)
+	if err != nil {
+		if err == bind.ErrNoCode {
+			// wait till contract is deployed
+			s.Require().Eventually(func() bool {
+				_, err = oracle.GetPriceData(&callOpts, checkDenom)
+				return err == nil
+			}, 2*time.Minute, 10*time.Second)
+		}
+	}
+
+	// eventually the contract will have price, deviation and median data
+	s.Require().Eventually(func() bool {
+		rate, err := oracle.GetPriceData(&callOpts, checkDenom)
+		s.Require().NoError(err)
+
+		deviationRate, err := session.GetDeviationData(checkDenom)
+		s.Require().NoError(err)
+
+		medianRate, err := session.GetMedianData(checkDenom)
+		s.Require().NoError(err)
+
+		if rate.Id.Int64() != 0 && deviationRate.Id.Int64() != 0 && medianRate.Id.Int64() != 0 {
+			return true
+		}
+
+		return false
+	}, 5*time.Minute, 10*time.Second)
+
+	// check individually for all assets
+	for _, asset := range mockPrices {
+		amount := relayer.DecTofactorBigInt(asset.Amount).Int64()
+		checkDenom := tools.StringToByte32(asset.Denom)
+
+		rate, err := oracle.GetPriceData(&callOpts, checkDenom)
+		s.Require().NoError(err)
+		s.Require().Equal(rate.Value.Int64(), amount)
+
+		deviationRate, err := session.GetDeviationData(checkDenom)
+		s.Require().NoError(err)
+		s.Require().Equal(deviationRate.Value.Int64(), amount)
+
+		medianRate, err := session.GetMedianData(checkDenom)
+		s.Require().NoError(err)
+		s.Require().Len(medianRate.Values, 1)
+		s.Require().Equal(medianRate.Values[0].Int64(), amount)
+	}
+}
+
+func (s *IntegrationTestSuite) TestQueryBulkRates() {
+	address := common.HexToAddress(orchestrator.ContractAddress)
+	ethClient, err := ethclient.Dial(s.orchestrator.EVMRpc)
+	s.Require().NoError(err)
+
+	oracle, err := client.NewOracle(address, ethClient)
+	s.Require().NoError(err)
+
+	mockPrices := s.priceServer.GetMockPrices()
+	s.Require().NotZero(len(mockPrices))
+
+	callOpts := bind.CallOpts{
+		Pending: false,
+	}
+
+	session := &client.OracleSession{
+		Contract: oracle,
+		CallOpts: callOpts,
+	}
+
+	// handle delay in deployment of contract
+	checkDenom := tools.StringToByte32(mockPrices[0].Denom)
+	_, err = oracle.GetPriceData(&callOpts, checkDenom)
+	if err != nil {
+		if err == bind.ErrNoCode {
```

Comment:
What happens when err != bind.ErrNoCode? It looks like it will continue on and assume GetPriceData returned prices.